pkgs/development/libraries/openmpi/default.nix

```diff
@@ -1,136 +1,270 @@
-    ++ lib.optionals fabricSupport [ "--with-psm2=${lib.getDev libpsm2}" "--with-libfabric=${lib.getDev libfabric}" ]
-    ;
+    (lib.withFeatureAs cudaSupport "cuda" (lib.getDev cudaPackages.cuda_cudart))
+    (lib.enableFeature cudaSupport "dlopen")
+    (lib.withFeatureAs fabricSupport "psm2" (lib.getDev libpsm2))
+    (lib.withFeatureAs fabricSupport "ofi" (lib.getDev libfabric))
+    # The flag --without-ofi-libdir is not supported from some reason, so we
+    # don't use lib.withFeatureAs
+  ] ++ lib.optionals fabricSupport [ "--with-ofi-libdir=${lib.getLib libfabric}/lib" ];
 
   enableParallelBuilding = true;
 
-  postInstall = ''
-    find $out/lib/ -name "*.la" -exec rm -f \{} \;
-
-    for f in mpi shmem osh; do
-      for i in f77 f90 CC c++ cxx cc fort; do
-        moveToOutput "bin/$f$i" "''${!outputDev}"
-        echo "move $fi$i"
-        moveToOutput "share/openmpi/$f$i-wrapper-data.txt" "''${!outputDev}"
-      done
-    done
-
-    for i in ortecc orte-info ompi_info oshmem_info opal_wrapper; do
-      moveToOutput "bin/$i" "''${!outputDev}"
-    done
-
-    moveToOutput "share/openmpi/ortecc-wrapper-data.txt" "''${!outputDev}"
-   '';
-
-  postFixup = ''
-    remove-references-to -t $dev $(readlink -f $out/lib/libopen-pal${stdenv.hostPlatform.extensions.sharedLibrary})
-    remove-references-to -t $man $(readlink -f $out/lib/libopen-pal${stdenv.hostPlatform.extensions.sharedLibrary})
-
-    # The path to the wrapper is hard coded in libopen-pal.so, which we just cleared.
-    wrapProgram $dev/bin/opal_wrapper \
-      --set OPAL_INCLUDEDIR $dev/include \
-      --set OPAL_PKGDATADIR $dev/share/openmpi
+  postInstall =
+    let
+      # The file names we need to iterate are a combination of ${p}${s}, and there
+      # are 7x3 such options. We use lib.mapCartesianProduct to iterate them all.
+      fileNamesToIterate = {
+        p = [
+          "mpi"
+          "shmem"
+          "osh"
+        ];
+        s =
+          [
+            "CC"
+            "c++"
+            "cxx"
+            "cc"
+          ]
+          ++ lib.optionals fortranSupport [
+            "f77"
+            "f90"
+            "fort"
+          ];
+      };
+      wrapperDataSubstitutions =
+        {
+          # The attr key is the filename prefix. The list's 1st value is the
+          # compiler=_ line that should be replaced by a compiler=#2 string, where
+          # #2 is the 2nd value in the list.
+          "cc" = [
+            "gcc"
+            "${targetPackages.stdenv.cc}/bin/${targetPackages.stdenv.cc.targetPrefix}cc"
+          ];
+          "c++" = [
+            "g++"
+            "${targetPackages.stdenv.cc}/bin/${targetPackages.stdenv.cc.targetPrefix}c++"
+          ];
+        }
+        // lib.optionalAttrs fortranSupport {
+          "fort" = [
+            "gfortran"
+            "${targetPackages.gfortran}/bin/${targetPackages.gfortran.targetPrefix}gfortran"
```

Comment:
What's the reason for using `targetPackages.gfortran` here? It'll only work on non-cross where `targetPackages.gfortran` is the same as `gfortran`? Asking for https://github.com/NixOS/nixpkgs/pull/341269

If you were just following the `targetPackages.stdenv` above then `targetPackages.stdenv` is the same offset as `gfortran`

```
nix-repl> pkgsCross.aarch64-multiplatform.gfortran.cc.stdenv.cc
«derivation /nix/store/cy0iv1sk4c60df358j3ff8xb2x0db1qh-aarch64-unknown-linux-gnu-gcc-wrapper-13.3.0.drv»

nix-repl> pkgsCross.aarch64-multiplatform.targetPackages.stdenv.cc
«derivation /nix/store/bh6d4yc7dyx1694gz3bx65l0flv1blc2-gcc-wrapper-13.3.0.drv»


nix-repl> pkgsCross.aarch64-multiplatform.targetPackages.stdenv.cc.stdenv.hostPlatform.system
"aarch64-linux"

nix-repl> pkgsCross.aarch64-multiplatform.targetPackages.stdenv.cc.stdenv.targetPlatform.system
"aarch64-linux"

nix-repl> pkgsCross.aarch64-multiplatform.gfortran.cc.stdenv.hostPlatform.system
"aarch64-linux"

nix-repl> pkgsCross.aarch64-multiplatform.gfortran.cc.stdenv.targetPlatform.system
"aarch64-linux"
```

Reply:
> EDIT: disregard the below, it should be for when `openmpi` goes in `nativeBuildInputs` right?

Correct. I'm not sure how to test that with `nix-repl` like tests as you tried to demonstrate, but yes.